File: sdk/go/provider/runtime/lifecycle.go
// Package runtimeprovider binds provider-owned runtime host policy to the
// canonical runtime lifecycle model.
package runtimeprovider

import (
	"context"
	"encoding/json"
	"errors"
	"strings"

	runtimesdk "easynet.run/cli/sdk/go"
)

// Mode is the runtime host deployment role understood by the host provider.
type Mode string

const (
	ModeEdge      Mode = "device"
	ModeAuthority Mode = "hub"
	ModeCombined  Mode = "both"
)

// RuntimeHostStartConfig is the provider-owned runtime host start policy.
type RuntimeHostStartConfig struct {
	Mode              Mode              `json:"mode"`
	Realm             string            `json:"realm,omitempty"`
	RuntimeInstanceID string            `json:"runtime_instance_id,omitempty"`
	HomeDir           string            `json:"home_dir,omitempty"`
	RuntimeBin        string            `json:"runtime_bin,omitempty"`
	WorkingDir        string            `json:"working_dir,omitempty"`
	LogPath           string            `json:"log_path,omitempty"`
	Detached          bool              `json:"detached,omitempty"`
	Env               map[string]string `json:"env,omitempty"`
	UDSPath           string            `json:"uds_path,omitempty"`
	ListenTCP         string            `json:"listen_tcp,omitempty"`
	TLSCertPath       string            `json:"tls_cert_path,omitempty"`
	TLSKeyPath        string            `json:"tls_key_path,omitempty"`
	AuthorityEndpoint string            `json:"authority_endpoint,omitempty"`
	TrustPath         string            `json:"trust_path,omitempty"`
}

// RuntimeHostAttachOptions identifies an existing runtime host.
type RuntimeHostAttachOptions = runtimesdk.RuntimeHostAttachOptions

// RuntimeHostDiscoverOptions controls runtime host discovery.
type RuntimeHostDiscoverOptions struct {
	ControlEndpoint string `json:"control_endpoint,omitempty"`
	ControlPath     string `json:"control_path,omitempty"`
	HomeDir         string `json:"home_dir,omitempty"`
}

// RuntimeHostStopOptions controls runtime host shutdown.
type RuntimeHostStopOptions = runtimesdk.RuntimeHostStopOptions

// ValidationError reports invalid runtime provider policy before transport use.
type ValidationError struct {
	message string
}

func (e *ValidationError) Error() string {
	if e == nil {
		return ""
	}
	return e.message
}

// Validate enforces runtime host deployment policy before transport use.
func (c RuntimeHostStartConfig) Validate() error {
	switch c.Mode {
	case ModeEdge:
		if strings.TrimSpace(c.ListenTCP) != "" {
			return &ValidationError{message: "edge runtime host mode must not accept a public TCP listener"}
		}
	case ModeAuthority, ModeCombined:
		if strings.TrimSpace(c.ListenTCP) != "" &&
			(strings.TrimSpace(c.TLSCertPath) == "" || strings.TrimSpace(c.TLSKeyPath) == "") {
			return &ValidationError{message: "public TCP listener requires TLS material"}
		}
	default:
		return &ValidationError{message: "mode must be edge, authority, or combined"}
	}
	return nil
}

// RuntimeHostStartPayload lowers validated runtime process policy for the
// canonical runtime-host lifecycle transport.
func (c RuntimeHostStartConfig) RuntimeHostStartPayload() ([]byte, error) {
	return json.Marshal(c)
}

// RuntimeHostDiscoverPayload lowers runtime host discovery policy.
func (o RuntimeHostDiscoverOptions) RuntimeHostDiscoverPayload() ([]byte, error) {
	return json.Marshal(o)
}

// IsValidationError reports whether an error came from runtime provider policy.
func IsValidationError(err error) bool {
	var target *ValidationError
	return errors.As(err, &target)
}

// Lifecycle is the runtime provider facade over one canonical RuntimeHost.
// It owns no lifecycle state and performs no implicit provider selection.
type Lifecycle struct {
	host *runtimesdk.RuntimeHost
}

// NewLifecycle binds an explicit runtime lifecycle transport to the canonical host.
func NewLifecycle(transport runtimesdk.RuntimeLifecycleTransport) (*Lifecycle, error) {
	host, err := runtimesdk.NewRuntimeHost(transport)
	if err != nil {
		return nil, err
	}
	return &Lifecycle{host: host}, nil
}

func (l *Lifecycle) Discover(ctx context.Context, opts RuntimeHostDiscoverOptions) (runtimesdk.RuntimeHostEndpoints, error) {
	return l.host.DiscoverRuntime(ctx, opts)
}

func (l *Lifecycle) Start(ctx context.Context, cfg RuntimeHostStartConfig) (*runtimesdk.RuntimeHandle, error) {
	return l.host.StartRuntime(ctx, cfg)
}

func (l *Lifecycle) Attach(ctx context.Context, opts RuntimeHostAttachOptions) (*runtimesdk.RuntimeHandle, error) {
	return l.host.AttachRuntime(ctx, opts)
}

func (l *Lifecycle) ConnectLocal(ctx context.Context, opts runtimesdk.ConnectOptions) (*runtimesdk.RuntimeClient, error) {
	return l.host.ConnectLocal(ctx, opts)
}

File: sdk/python/easynet_sdk/providers/runtime/lifecycle.py
"""Runtime provider lifecycle policy and stable host DTOs."""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from enum import StrEnum
from typing import Mapping

from ...errors import ErrorCode, RetryHint, SDKError


class RuntimeHostMode(StrEnum):
    """Runtime host deployment role understood by the local host provider."""

    EDGE = "device"
    AUTHORITY = "hub"
    COMBINED = "both"


@dataclass(frozen=True)
class RuntimeHostStartConfig:
    """Runtime host process start policy."""

    mode: RuntimeHostMode
    realm: str = ""
    runtime_instance_id: str = ""
    home_dir: str = ""
    runtime_bin: str = ""
    log_path: str = ""
    detached: bool = False
    env: Mapping[str, str] = field(default_factory=dict)
    uds_path: str = ""
    listen_tcp: str = ""
    tls_cert_path: str = ""
    tls_key_path: str = ""
    authority_endpoint: str = ""
    trust_path: str = ""

    def validate(self) -> None:
        if self.mode == RuntimeHostMode.EDGE and self.listen_tcp.strip():
            raise _invalid_lifecycle(
                "edge runtime host mode must not accept a public TCP listener"
            )
        if (
            self.mode in {RuntimeHostMode.AUTHORITY, RuntimeHostMode.COMBINED}
            and self.listen_tcp.strip()
            and (not self.tls_cert_path.strip() or not self.tls_key_path.strip())
        ):
            raise _invalid_lifecycle("public TCP listener requires TLS material")

    def to_json_dict(self) -> dict[str, object]:
        value: dict[str, object] = {"mode": self.mode.value}
        for key in (
            "realm",
            "runtime_instance_id",
            "home_dir",
            "runtime_bin",
            "log_path",
            "uds_path",
            "listen_tcp",
            "tls_cert_path",
            "tls_key_path",
            "authority_endpoint",
            "trust_path",
        ):
            item = getattr(self, key)
            if item:
                value[key] = item
        if self.detached:
            value["detached"] = True
        if self.env:
            value["env"] = dict(self.env)
        return value

    def to_json_bytes(self) -> bytes:
        return json.dumps(
            self.to_json_dict(), separators=(",", ":"), sort_keys=True
        ).encode("utf-8")


@dataclass(frozen=True)
class RuntimeHostDiscoverConfig:
    """Runtime host provider discovery policy, including the host state root."""

    control_endpoint: str = ""
    control_path: str = ""
    home_dir: str = ""

    def to_json_bytes(self) -> bytes:
        return _json_bytes(
            {
                "control_endpoint": self.control_endpoint,
                "control_path": self.control_path,
                "home_dir": self.home_dir,
            }
        )


def _invalid_lifecycle(message: str) -> SDKError:
    return SDKError(
        code=ErrorCode.INVALID_ARGUMENT,
        stage="sdk",
        retry=RetryHint.NEVER,
        retryable=False,
        message=message,
    )


def _json_bytes(value: Mapping[str, object]) -> bytes:
    compact = {key: item for key, item in value.items() if item not in ("", 0, False)}
    return json.dumps(compact, separators=(",", ":"), sort_keys=True).encode("utf-8")
